In [67]:
#importaciones de datos
import pandas as pd

#importación de src codigo de soporte con las funciones necesarias para la limpieza
import sys
sys.path.append("..")
from src import sp_eda_limpieza as sp

In [68]:
#ajustes de visualización
pd.set_option('display.max_columns', None)

In [69]:
#lectura de los ficheros de datos
df_bankadditional_raw=pd.read_csv(r'../data/fuentes de datos/bank-additional.csv')

In [70]:
sheets=pd.read_excel(r'../data/fuentes de datos/customer-details.xlsx',sheet_name=None)
df_customerdetails_raw=pd.concat(sheets.values(),ignore_index=True)

In [71]:
#se cran las variables que almacenan la copia del fichero original. estas se manipularán para ser limpiadas
df_bankadditional=df_bankadditional_raw.copy()
df_customerdetails=df_customerdetails_raw.copy()

In [72]:
#Eda preeliminar del fichero customer details
sp.eda_preliminar(df_customerdetails)

Income  Kidhome  Teenhome Dt_Customer  NumWebVisitsMonth  \
39175   25086        1         1  2014-01-02                  2   
11795   21147        2         1  2012-04-15                 31   
11575  176094        0         1  2012-01-16                 21   
33929  177079        2         0  2014-11-08                  3   
22211   32335        0         0  2013-03-06                  8   

                                         ID  
39175  e94d5c90-8988-4b79-bdde-97af17721f1c  
11795  50ef7511-1f89-468a-b13f-b9cb2c7c33b3  
11575  12bffae6-9288-4a9c-a805-5c1248ad0d2e  
33929  c65e3d9e-9d62-4c4e-b9c7-ba1b43ec8537  
22211  b8cc0958-9569-4311-a365-2bd6408e3816

--------------------
INFO
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 43170 entries, 0 to 43169
Data columns (total 6 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   Income             43170 non-null  int64         
 1   Kidhome            43170 non-null  int64         
 2   Teenhome           43170 non-null  int64         
 3   Dt_Customer        43170 non-null  datetime64[ns]
 4   NumWebVisitsMonth  43170 non-null  int64         
 5   ID                 43170 non-null  object        
dtypes: datetime64[ns](1), int64(4), object(1)
memory usage: 2.0+ MB


None

--------------------
NULOS


Income               0.0
Kidhome              0.0
Teenhome             0.0
Dt_Customer          0.0
NumWebVisitsMonth    0.0
ID                   0.0
dtype: float64

--------------------
duplicados
0
VALUE COUNT


ID
9727bc06-c11a-461a-a5bb-3d210467cc2a    1
089b39d8-e4d0-461b-87d4-814d71e0e079    1
5006b75e-18bb-46e9-ae1a-89f7eb79cc8a    1
e554d2ad-6c83-4603-93d4-fc596b807eff    1
3959015a-9713-4877-9a43-740e34c29eaf    1
                                       ..
5e3483e5-236d-437d-8351-541f9d09b9dd    1
d63ede72-0b6d-45b1-8872-385ac6897f65    1
eca60b76-70b6-4077-80ba-bc52e8ebb0eb    1
9991fafb-4447-451a-8be2-b0df6098d13e    1
3f9f49b5-e410-4948-bf6e-f9244f04918b    1
Name: count, Length: 43170, dtype: int64

--------------------
VALORES NUMÉRICOS


count                           mean  \
Income             43170.0                   93227.389622   
Kidhome            43170.0                       1.004749   
Teenhome           43170.0                       0.998633   
Dt_Customer          43170  2013-05-11 22:09:47.407922176   
NumWebVisitsMonth  43170.0                       16.59129   

                                   min                  25%  \
Income                          5841.0              49608.0   
Kidhome                            0.0                  0.0   
Teenhome                           0.0                  0.0   
Dt_Customer        2012-01-01 00:00:00  2012-07-16 00:00:00   
NumWebVisitsMonth                  1.0                  9.0   

                                   50%                  75%  \
Income                         93009.5             136740.5   
Kidhome                            1.0                  2.0   
Teenhome                           1.0                  2.0   
Dt_Customer        2013-03-01 00:00:00  2014-03-28 00:00:00   
NumWebVisitsMonth                 17.0                 25.0   

                                   max           std  
Income                        180802.0  50498.181989  
Kidhome                            2.0      0.815996  
Teenhome                           2.0      0.816008  
Dt_Customer        2014-12-31 00:00:00           NaN  
NumWebVisitsMonth                 32.0      9.241769

## Análisis
- Borrar columna unnamed:0
- nulos 0
- duplicados 0
- categorías: todo ok
## To do
- Borrar columa unnamed:0

In [73]:
sp.borrar_unnamed(df_customerdetails)

Columnas originales: Index(['Income', 'Kidhome', 'Teenhome', 'Dt_Customer', 'NumWebVisitsMonth',
       'ID'],
      dtype='object')
La columna 'Unnamed: 0' no se encuentra en el DataFrame


Income  Kidhome  Teenhome Dt_Customer  NumWebVisitsMonth  \
1543  124002        1         2  2012-03-16                 16   

                                        ID  
1543  a20b899a-9824-4f54-b203-035d34a4354a

In [74]:
#Eda preeliminar del fichero bankadditional
sp.eda_preliminar(df_bankadditional)

Unnamed: 0   age          job  marital          education  default  \
30336       30480  51.0  blue-collar  MARRIED           basic.4y      0.0   
37752       37919  26.0       admin.   SINGLE  university.degree      0.0   
27632       27776  32.0   management   SINGLE  university.degree      0.0   
33173       33317  41.0   management  MARRIED        high.school      NaN   
2459         2556   NaN   management  MARRIED  university.degree      0.0   

       housing  loan    contact  duration  campaign  pdays  previous  \
30336      1.0   0.0   cellular        59         2    999         0   
37752      1.0   0.0  telephone       471         1      7         1   
27632      0.0   0.0   cellular        93         2    999         1   
33173      1.0   0.0   cellular       527         2    999         0   
2459       0.0   0.0  telephone       129         4    999         0   

          poutcome  emp.var.rate cons.price.idx cons.conf.idx euribor3m  \
30336  NONEXISTENT          -1.8         92,893         -46,2       NaN   
37752      SUCCESS          -3.4         92,379         -29,8     0,809   
27632      FAILURE          -1.8         92,843           -50     1,703   
33173  NONEXISTENT          -1.8         92,893         -46,2     1,291   
2459   NONEXISTENT           1.1         93,994         -36,4     4,856   

      nr.employed    y               date  latitude  longitude  \
30336      5099,1   no  10-diciembre-2017    33.662    -76.065   
37752      5017,5  yes      26-abril-2019    39.367    -69.760   
27632      5099,1   no      13-marzo-2018    27.795    -79.417   
33173      5099,1   no       21-mayo-2015    33.470    -82.324   
2459         5191   no       19-mayo-2015    38.536    -96.518   

                                        id_  
30336  42e7bf3c-54ce-49f3-af5d-c57106f9e1bc  
37752  19b2a2c8-a68f-4ce3-9d5b-8989031eb8bd  
27632  4561e3c6-2876-4c3b-9dea-cadac11f584b  
33173  eafd6b08-0c4e-48f2-909e-3cb6199cfc44  
2459   15a55050-6271-47fe-bdb1-9feaf8ece0c6

--------------------
INFO
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 43000 entries, 0 to 42999
Data columns (total 24 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Unnamed: 0      43000 non-null  int64  
 1   age             37880 non-null  float64
 2   job             42655 non-null  object 
 3   marital         42915 non-null  object 
 4   education       41193 non-null  object 
 5   default         34019 non-null  float64
 6   housing         41974 non-null  float64
 7   loan            41974 non-null  float64
 8   contact         43000 non-null  object 
 9   duration        43000 non-null  int64  
 10  campaign        43000 non-null  int64  
 11  pdays           43000 non-null  int64  
 12  previous        43000 non-null  int64  
 13  poutcome        43000 non-null  object 
 14  emp.var.rate    43000 non-null  float64
 15  cons.price.idx  42529 non-null  object 
 16  cons.conf.idx   43000 non-null  object 
 17  eurib

None

--------------------
NULOS


Unnamed: 0         0.00
age               11.91
job                0.80
marital            0.20
education          4.20
default           20.89
housing            2.39
loan               2.39
contact            0.00
duration           0.00
campaign           0.00
pdays              0.00
previous           0.00
poutcome           0.00
emp.var.rate       0.00
cons.price.idx     1.10
cons.conf.idx      0.00
euribor3m         21.53
nr.employed        0.00
y                  0.00
date               0.58
latitude           0.00
longitude          0.00
id_                0.00
dtype: float64

--------------------
duplicados
0
VALUE COUNT


job
admin.           10873
blue-collar       9654
technician        7026
services          4162
management        3050
retired           1790
entrepreneur      1522
self-employed     1489
housemaid         1123
unemployed        1063
student            903
Name: count, dtype: int64

--------------------


marital
MARRIED     25999
SINGLE      12105
DIVORCED     4811
Name: count, dtype: int64

--------------------


education
university.degree      12722
high.school             9925
basic.9y                6309
professional.course     5477
basic.4y                4356
basic.6y                2386
illiterate                18
Name: count, dtype: int64

--------------------


contact
cellular     27396
telephone    15604
Name: count, dtype: int64

--------------------


poutcome
NONEXISTENT    37103
FAILURE         4461
SUCCESS         1436
Name: count, dtype: int64

--------------------


cons.price.idx
93,994    7938
93,918    6937
92,893    5985
93,444    5349
94,465    4522
93,2      3731
93,075    2552
92,201     803
92,963     742
92,431     471
92,649     375
94,215     320
94,199     316
92,843     298
92,379     274
93,369     271
94,027     242
94,055     236
93,876     224
94,601     212
92,469     184
92,713     179
93,749     179
94,767     130
93,798      48
92,756      11
Name: count, dtype: int64

--------------------


cons.conf.idx
-36,4    8020
-42,7    7004
-46,2    6057
-36,1    5408
-41,8    4568
-42      3782
-47,1    2581
-31,4     815
-40,8     748
-26,9     477
-30,1     379
-40,3     328
-37,5     318
-50       299
-29,8     277
-34,8     275
-38,3     245
-39,8     239
-40       225
-49,5     214
-33,6     185
-34,6     183
-33       181
-50,8     133
-40,4      48
-45,9      11
Name: count, dtype: int64

--------------------


euribor3m
4,857    2287
4,962    2124
4,963    2019
4,961    1594
4,964     964
         ... 
5,045       1
4,921       1
0,895       1
0,956       1
0,953       1
Name: count, Length: 309, dtype: int64

--------------------


nr.employed
5228,1    16980
5099,1     8937
5191       8020
5195,8     3830
5076,2     1748
5017,5     1133
4991,6      812
5008,7      683
4963,6      665
5023,5      181
5176,3       11
Name: count, dtype: int64

--------------------


y
no     38156
yes     4844
Name: count, dtype: int64

--------------------


date
28-febrero-2018      95
28-febrero-2016      94
28-febrero-2015      92
28-febrero-2017      89
28-febrero-2019      85
                     ..
17-diciembre-2018    11
21-diciembre-2019    10
14-abril-2015        10
4-abril-2017         10
20-noviembre-2015    10
Name: count, Length: 1825, dtype: int64

--------------------


id_
993bbbd6-4dbc-4a40-a408-f91f8462bee6    1
089b39d8-e4d0-461b-87d4-814d71e0e079    1
e9d37224-cb6f-4942-98d7-46672963d097    1
3f9f49b5-e410-4948-bf6e-f9244f04918b    1
9991fafb-4447-451a-8be2-b0df6098d13e    1
                                       ..
66a752e2-b2f1-440c-9a8f-cc3b10d74dd0    1
ca5c0d7a-8bbd-42b3-8351-1074c5be011a    1
8d700df5-0c33-4517-8cf8-92e1c92c9c12    1
ea6b7d04-9271-4c0a-a01f-07795d164aba    1
87b79988-2be5-419d-88f4-56655852c565    1
Name: count, Length: 43000, dtype: int64

--------------------
VALORES NUMÉRICOS


count          mean           std      min          25%  \
Unnamed: 0    43000.0  20651.099047  11868.770913    0.000  10381.75000   
age           37880.0     39.977112     10.437957   17.000     32.00000   
default       34019.0      0.000088      0.009390    0.000      0.00000   
housing       41974.0      0.535998      0.498708    0.000      0.00000   
loan          41974.0      0.155620      0.362499    0.000      0.00000   
duration      43000.0    257.739279    258.666033    0.000    102.00000   
campaign      43000.0      2.567233      2.772294    1.000      1.00000   
pdays         43000.0    962.330953    187.260394    0.000    999.00000   
previous      43000.0      0.174023      0.497366    0.000      0.00000   
emp.var.rate  43000.0      0.077128      1.573898   -3.400     -1.80000   
latitude      43000.0     36.856697      7.225948   24.396     30.61475   
longitude     43000.0    -95.939067     16.752282 -124.997   -110.49425   

                     50%          75%        max  
Unnamed: 0    20642.5000  30930.25000  41187.000  
age              38.0000     47.00000     98.000  
default           0.0000      0.00000      1.000  
housing           1.0000      1.00000      1.000  
loan              0.0000      0.00000      1.000  
duration        179.0000    319.00000   4918.000  
campaign          2.0000      3.00000     56.000  
pdays           999.0000    999.00000    999.000  
previous          0.0000      0.00000      7.000  
emp.var.rate      1.1000      1.40000      1.400  
latitude         36.7610     43.11325     49.384  
longitude       -95.8995    -81.42775    -66.937

## Análisis to do 
- Borrar columna unnamed
- Poner iniciales de las columnas en minúsculas
- conversión de booleanos a sí no: default,housing,loan
- conversión de valores a int: nr.employed
- conversión de valores float: cons.price.idx,cons.conf.ids, euribor3m
- eliminar pdays: el 75% de lso valores son 999
- pasar  a fecha la columna date
- pasar valores a minúsculas: marital, poutcome
- education, cambiar el punto por una _
- cambiar - por _ en job
- cambiar Id_ por ID
- Cambiar Y por Subscribed
- Unir los dos dataframes
- gestión de nulos
- conversión de coordenadas a ciudad, región, país

In [75]:
#borrar la columna unnamed
sp.borrar_unnamed(df_bankadditional)

Columnas originales: Index(['Unnamed: 0', 'age', 'job', 'marital', 'education', 'default',
       'housing', 'loan', 'contact', 'duration', 'campaign', 'pdays',
       'previous', 'poutcome', 'emp.var.rate', 'cons.price.idx',
       'cons.conf.idx', 'euribor3m', 'nr.employed', 'y', 'date', 'latitude',
       'longitude', 'id_'],
      dtype='object')
Columna 'Unnamed: 0' eliminada


age          job  marital education  default  housing  loan   contact  \
25135  40.0  blue-collar  MARRIED  basic.9y      NaN      0.0   0.0  cellular   

       duration  campaign  pdays  previous     poutcome  emp.var.rate  \
25135       113         3    999         0  NONEXISTENT          -0.1   

      cons.price.idx cons.conf.idx euribor3m nr.employed   y           date  \
25135           93,2           -42     4,153      5195,8  no  27-abril-2018   

       latitude  longitude                                   id_  
25135    46.029    -97.652  9ca9cdb7-972c-4c8e-a4c3-d986532ac9c3

In [76]:
#pasar columnas a mayúscula
sp.columnas_mayuscula(df_bankadditional)

Columnas pasadas a mayúscualas


Age            Job  Marital          Education  Default  Housing  \
20338  50.0  self-employed  MARRIED  university.degree      0.0      1.0   

       Loan   Contact  Duration  Campaign  Pdays  Previous     Poutcome  \
20338   0.0  cellular       155         1    999         0  NONEXISTENT   

       Emp.var.rate Cons.price.idx Cons.conf.idx Euribor3m Nr.employed   Y  \
20338           1.4         93,444         -36,1       NaN      5228,1  no   

                Date  Latitude  Longitude  \
20338  31-marzo-2019    29.048    -93.449   

                                        Id_  
20338  21da6970-82b5-46bb-b99e-dd4cd5b12d29

In [77]:
#transformar valores 1 / 0 a yes, no
sp.transformar_booleanos(df_bankadditional)

Columnas booleanas detectadas: ['Default', 'Housing', 'Loan']

Muestra del DataFrame:


Age            Job  Marital            Education Default Housing Loan  \
32745  53.0  self-employed  MARRIED             basic.4y     NaN     yes   no   
4702   33.0    blue-collar  MARRIED  professional.course      no     yes   no   
23966   NaN         admin.  MARRIED             basic.9y      no     yes   no   
5497   34.0         admin.  MARRIED    university.degree      no      no  yes   
20756   NaN         admin.  MARRIED    university.degree      no     yes   no   

         Contact  Duration  Campaign  Pdays  Previous     Poutcome  \
32745   cellular       145         2    999         1      FAILURE   
4702   telephone       161         4    999         0  NONEXISTENT   
23966  telephone       453         1    999         0  NONEXISTENT   
5497   telephone       165         2    999         0  NONEXISTENT   
20756   cellular       123         1    999         0  NONEXISTENT   

       Emp.var.rate Cons.price.idx Cons.conf.idx Euribor3m Nr.employed    Y  \
32745          -1.8         92,893         -46,2     1,299      5099,1   no   
4702            1.1         93,994         -36,4     4,858        5191   no   
23966          -0.1           93,2           -42     4,474      5195,8  yes   
5497            1.1         93,994         -36,4     4,857        5191   no   
20756           1.4         93,444         -36,1     4,964      5228,1   no   

                     Date  Latitude  Longitude  \
32745  16-septiembre-2016    38.662   -118.490   
4702   16-septiembre-2015    29.770    -98.213   
23966      2-octubre-2015    25.044    -94.594   
5497    18-noviembre-2019    42.785   -102.596   
20756      25-agosto-2016    26.175   -112.426   

                                        Id_  
32745  d9feecbc-614a-461c-8891-ce973b974303  
4702   d6093e3f-a8f8-4d28-b637-131531575855  
23966  f623183c-853c-4d09-bc43-61cc5c3993c6  
5497   e6bf26a5-ee51-40f7-8745-75153f9bf045  
20756  aadb167d-57b1-4833-8e72-9a467c412509

In [78]:
#cambiar , por . y convertir a float
sp.convertir_a_float(df_bankadditional)

Se han modificado las siguientes columnas:['Cons.price.idx', 'Cons.conf.idx', 'Euribor3m', 'Nr.employed']


Age          Job Marital    Education Default Housing Loan   Contact  \
35348  45.0  blue-collar  SINGLE  high.school     NaN     yes   no  cellular   

       Duration  Campaign  Pdays  Previous Poutcome  Emp.var.rate  \
35348        29         6    999         1  FAILURE          -1.8   

       Cons.price.idx  Cons.conf.idx  Euribor3m  Nr.employed   Y  \
35348          92.893          -46.2      1.244       5099.1  no   

                Date  Latitude  Longitude  \
35348  20-marzo-2019    27.861    -87.019   

                                        Id_  
35348  6fdd8be3-fadd-4643-82eb-d44a4b38af3f

In [79]:
#redondear decimal y convertir a int
sp.float_a_int(df_bankadditional,'Nr.employed')

La columna Nr.employed ha sido cambiada a numero entero y redondeada
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 43000 entries, 0 to 42999
Data columns (total 23 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Age             37880 non-null  float64
 1   Job             42655 non-null  object 
 2   Marital         42915 non-null  object 
 3   Education       41193 non-null  object 
 4   Default         34019 non-null  object 
 5   Housing         41974 non-null  object 
 6   Loan            41974 non-null  object 
 7   Contact         43000 non-null  object 
 8   Duration        43000 non-null  int64  
 9   Campaign        43000 non-null  int64  
 10  Pdays           43000 non-null  int64  
 11  Previous        43000 non-null  int64  
 12  Poutcome        43000 non-null  object 
 13  Emp.var.rate    43000 non-null  float64
 14  Cons.price.idx  42529 non-null  float64
 15  Cons.conf.idx   43000 non-null  float64
 16  Euribor

None

Age          Job  Marital Education Default Housing Loan   Contact  \
36474  53.0  blue-collar  MARRIED  basic.6y      no      no   no  cellular   

       Duration  Campaign  Pdays  Previous Poutcome  Emp.var.rate  \
36474       592         3    999         1  FAILURE          -2.9   

       Cons.price.idx  Cons.conf.idx  Euribor3m  Nr.employed    Y  \
36474          92.963          -40.8       1.26         5076  yes   

               Date  Latitude  Longitude                                   Id_  
36474  6-junio-2016    27.824    -87.791  4407de3a-edc5-4323-ab75-39bd58d41c0e

In [80]:
sp.eliminar_columna(df_bankadditional,'Pdays')

Se ha eliminado la columna Pdays


Age     Job Marital          Education Default Housing Loan   Contact  \
40426  NaN  admin.  SINGLE  university.degree      no      no   no  cellular   

       Duration  Campaign  Previous     Poutcome  Emp.var.rate  \
40426       154         1         0  NONEXISTENT          -1.1   

       Cons.price.idx  Cons.conf.idx  Euribor3m  Nr.employed   Y  \
40426          94.199          -37.5      0.884         4964  no   

                  Date  Latitude  Longitude  \
40426  28-febrero-2019    33.631   -115.019   

                                        Id_  
40426  caf009e4-d9f8-487c-9fe3-446a13ace486

In [81]:
#convertir la columna a fecha
sp.convertir_fecha(df_bankadditional,'Date','mixed')

Se ha convertido la columna Date a fecha
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 43000 entries, 0 to 42999
Data columns (total 22 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   Age             37880 non-null  float64       
 1   Job             42655 non-null  object        
 2   Marital         42915 non-null  object        
 3   Education       41193 non-null  object        
 4   Default         34019 non-null  object        
 5   Housing         41974 non-null  object        
 6   Loan            41974 non-null  object        
 7   Contact         43000 non-null  object        
 8   Duration        43000 non-null  int64         
 9   Campaign        43000 non-null  int64         
 10  Previous        43000 non-null  int64         
 11  Poutcome        43000 non-null  object        
 12  Emp.var.rate    43000 non-null  float64       
 13  Cons.price.idx  42529 non-null  float64       
 14  Cons.conf.idx

Age       Job  Marital    Education Default Housing Loan   Contact  \
25881  34.0  services  MARRIED  high.school      no      no   no  cellular   

       Duration  Campaign  Previous Poutcome  Emp.var.rate  Cons.price.idx  \
25881        89         3         1  FAILURE          -0.1            93.2   

       Cons.conf.idx  Euribor3m  Nr.employed   Y       Date  Latitude  \
25881          -42.0       4.12         5196  no 2015-02-28    47.985   

       Longitude                                   Id_  
25881   -114.947  55873b46-7451-469e-afee-dcb505a379b6

In [82]:
#Convertir marital a minúsculas
sp.convertir_minusculas(df_bankadditional,'Marital')

Age         Job Marital          Education Default Housing Loan  \
40134  31.0  technician  single  university.degree      no      no   no   

         Contact  Duration  Campaign  Previous Poutcome  Emp.var.rate  \
40134  telephone       270         1         1  SUCCESS          -1.7   

       Cons.price.idx  Cons.conf.idx  Euribor3m  Nr.employed    Y       Date  \
40134          94.215          -40.3      0.899         4992  yes 2017-02-25   

       Latitude  Longitude                                   Id_  
40134    32.443    -91.304  2c961c3a-3357-46a0-bab3-a967978ac1e8

In [83]:
#convertir valores a minúsculas en la columna poutcome
sp.convertir_minusculas(df_bankadditional,'Poutcome')

Age          Job  Marital Education Default Housing Loan   Contact  \
31860  37.0  blue-collar  married  basic.6y     NaN      no   no  cellular   

       Duration  Campaign  Previous     Poutcome  Emp.var.rate  \
31860       635         2         0  nonexistent          -1.8   

       Cons.price.idx  Cons.conf.idx  Euribor3m  Nr.employed   Y       Date  \
31860          92.893          -46.2      1.327         5099  no 2017-11-16   

       Latitude  Longitude                                   Id_  
31860    33.424    -83.999  88e0900d-2d68-4ffb-b361-f584ca48500c

In [84]:
#CAMBIAR . POR _
sp.punto_barrabaja(df_bankadditional,'Education')

Se ha cambiado '.' por '_' en la columna Education


Age         Job Marital    Education Default Housing Loan   Contact  \
38115  28.0  technician  single  high_school      no      no   no  cellular   

       Duration  Campaign  Previous Poutcome  Emp.var.rate  Cons.price.idx  \
38115       216         1         1  failure          -3.4          92.431   

       Cons.conf.idx  Euribor3m  Nr.employed    Y       Date  Latitude  \
38115          -26.9      0.742         5018  yes 2016-05-22    27.977   

       Longitude                                   Id_  
38115   -116.371  95e33bc6-f2a8-45c7-9868-ee1264e64eef

In [85]:
#Cambiar - por _
sp.guion_barrabaja(df_bankadditional,'Job')

Se ha cambiado '-' por '_' en la columna Job


Job
admin.           10873
blue_collar       9654
technician        7026
services          4162
management        3050
retired           1790
entrepreneur      1522
self_employed     1489
housemaid         1123
unemployed        1063
student            903
nan                345
Name: count, dtype: int64

In [86]:
#Cambiar el nombre de la columna Id_ por ID
sp.cambiar_columna(df_bankadditional,'Id_','ID')

Se ha cambiado el nombre de la columna Id_ por ID


Age     Job  Marital    Education Default Housing Loan   Contact  \
31996  38.0  admin.  married  high_school      no     yes   no  cellular   

       Duration  Campaign  Previous     Poutcome  Emp.var.rate  \
31996       213         5         0  nonexistent          -1.8   

       Cons.price.idx  Cons.conf.idx  Euribor3m  Nr.employed   Y       Date  \
31996          92.893          -46.2      1.313         5099  no 2018-07-01   

       Latitude  Longitude                                    ID  
31996    26.439   -113.997  06e009f9-8c78-44fd-ba81-b803ead8efb4

In [87]:
#Cambiar el nombre de la columna Y por Subscribed
sp.cambiar_columna(df_bankadditional,'Y','Subscribed')

Se ha cambiado el nombre de la columna Y por Subscribed


Age     Job   Marital          Education Default Housing Loan  \
22644  41.0  admin.  divorced  university_degree      no      no   no   

        Contact  Duration  Campaign  Previous     Poutcome  Emp.var.rate  \
22644  cellular       115         4         0  nonexistent           1.4   

       Cons.price.idx  Cons.conf.idx  Euribor3m  Nr.employed Subscribed  \
22644          93.444          -36.1      4.965         5228         no   

            Date  Latitude  Longitude                                    ID  
22644 2016-01-06    41.493   -102.516  64cda1e7-a2a0-472e-a661-e75e12eb9182

In [88]:
df_bankadditional.columns

Index(['Age', 'Job', 'Marital', 'Education', 'Default', 'Housing', 'Loan',
       'Contact', 'Duration', 'Campaign', 'Previous', 'Poutcome',
       'Emp.var.rate', 'Cons.price.idx', 'Cons.conf.idx', 'Euribor3m',
       'Nr.employed', 'Subscribed', 'Date', 'Latitude', 'Longitude', 'ID'],
      dtype='object')

Se unen los dos dataframes, df_bank_limpio y df_customerdetails

In [89]:
df_bank_merge=df_bankadditional.merge(df_customerdetails,how='outer',on='ID' )

In [90]:
sp.eda_preliminar(df_bank_merge)

Age            Job  Marital          Education Default Housing Loan  \
10863  30.0    blue_collar   single        high_school      no     yes   no   
1156   31.0         admin.  married        high_school     NaN      no   no   
14595  52.0    blue_collar  married           basic_9y      no     yes  yes   
36424  30.0  self_employed   single  university_degree      no      no   no   
9109   42.0    blue_collar  married           basic_6y     NaN     yes   no   

         Contact  Duration  Campaign  Previous     Poutcome  Emp.var.rate  \
10863   cellular     186.0       6.0       0.0  nonexistent           1.4   
1156    cellular      82.0       3.0       0.0  nonexistent           1.4   
14595  telephone     241.0       2.0       0.0  nonexistent           1.4   
36424  telephone     510.0       5.0       0.0  nonexistent           1.1   
9109    cellular     458.0       3.0       0.0  nonexistent          -1.8   

       Cons.price.idx  Cons.conf.idx  Euribor3m  Nr.employed Subscribed  \
10863          93.918          -42.7      4.961       5228.0         no   
1156           93.918          -42.7        NaN       5228.0         no   
14595          94.465          -41.8      4.959       5228.0         no   
36424          93.994          -36.4      4.864       5191.0         no   
9109           92.893          -46.2      1.334       5099.0         no   

            Date  Latitude  Longitude                                    ID  \
10863        NaT    47.776    -80.747  40274c21-1d38-40b3-b982-0478228f03c1   
1156  2016-06-30    48.149    -86.741  07061f32-3ccd-44ea-a464-1023466adabb   
14595 2018-02-11    32.052    -95.147  56ab52d9-1141-40e6-bd9a-3a0fb608d4f7   
36424 2018-09-19    43.644   -114.872  d842084e-831a-4dfb-92b1-812b23f6b170   
9109  2016-03-11    28.426    -90.035  35bcb4d5-b27b-494f-b5eb-a1679f40f6fc   

       Income  Kidhome  Teenhome Dt_Customer  NumWebVisitsMonth  
10863  140062        2         2  2012-05-01                 30  
1156    11887        2         0  2012-03-08                  5  
14595  139557        0         2  2012-06-06                 12  
36424   72916        1         2  2012-08-09                 11  
9109   149411        1         0  2014-02-12                 31

--------------------
INFO
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 43170 entries, 0 to 43169
Data columns (total 27 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   Age                37880 non-null  float64       
 1   Job                43000 non-null  object        
 2   Marital            42915 non-null  object        
 3   Education          43000 non-null  object        
 4   Default            34019 non-null  object        
 5   Housing            41974 non-null  object        
 6   Loan               41974 non-null  object        
 7   Contact            43000 non-null  object        
 8   Duration           43000 non-null  float64       
 9   Campaign           43000 non-null  float64       
 10  Previous           43000 non-null  float64       
 11  Poutcome           43000 non-null  object        
 12  Emp.var.rate       43000 non-null  float64       
 13  Cons.price.idx     42529 non-null  

None

--------------------
NULOS


Age                  12.25
Job                   0.39
Marital               0.59
Education             0.39
Default              21.20
Housing               2.77
Loan                  2.77
Contact               0.39
Duration              0.39
Campaign              0.39
Previous              0.39
Poutcome              0.39
Emp.var.rate          0.39
Cons.price.idx        1.48
Cons.conf.idx         0.39
Euribor3m            21.83
Nr.employed           0.39
Subscribed            0.39
Date                  0.97
Latitude              0.39
Longitude             0.39
ID                    0.00
Income                0.00
Kidhome               0.00
Teenhome              0.00
Dt_Customer           0.00
NumWebVisitsMonth     0.00
dtype: float64

--------------------
duplicados
0
VALUE COUNT


Job
admin.           10873
blue_collar       9654
technician        7026
services          4162
management        3050
retired           1790
entrepreneur      1522
self_employed     1489
housemaid         1123
unemployed        1063
student            903
nan                345
Name: count, dtype: int64

--------------------


Marital
married     25999
single      12105
divorced     4811
Name: count, dtype: int64

--------------------


Education
university_degree      12722
high_school             9925
basic_9y                6309
professional_course     5477
basic_4y                4356
basic_6y                2386
nan                     1807
illiterate                18
Name: count, dtype: int64

--------------------


Default
no     34016
yes        3
Name: count, dtype: int64

--------------------


Housing
yes    22498
no     19476
Name: count, dtype: int64

--------------------


Loan
no     35442
yes     6532
Name: count, dtype: int64

--------------------


Contact
cellular     27396
telephone    15604
Name: count, dtype: int64

--------------------


Poutcome
nonexistent    37103
failure         4461
success         1436
Name: count, dtype: int64

--------------------


Subscribed
no     38156
yes     4844
Name: count, dtype: int64

--------------------


ID
fffe84df-4e63-4465-82de-0d9821d919bf    1
0000e811-006e-4404-b535-89bf6cd96553    1
ffdbe65e-3249-456c-a8f9-544935b40fdd    1
ffdbe151-4f82-4c19-b63c-841c91fc3084    1
ffda2b4b-0b68-493e-ac68-f81614c57f5d    1
                                       ..
0006c8fd-9dea-439d-a3b9-571b3901ff58    1
000625bd-cc08-4653-b8b7-1853c458527d    1
0004e1d1-958d-4abf-a57c-9b9c7be887a0    1
00024507-c59b-4eee-86d5-cc341b96eb6d    1
000165f9-20c0-4cb5-bd47-6233b92655c1    1
Name: count, Length: 43170, dtype: int64

--------------------
VALORES NUMÉRICOS


count                           mean  \
Age                37880.0                      39.977112   
Duration           43000.0                     257.739279   
Campaign           43000.0                       2.567233   
Previous           43000.0                       0.174023   
Emp.var.rate       43000.0                       0.077128   
Cons.price.idx     42529.0                      93.574219   
Cons.conf.idx      43000.0                     -40.509049   
Euribor3m          33744.0                       3.616521   
Nr.employed        43000.0                    5166.835721   
Date                 42752  2017-07-01 08:57:44.595808256   
Latitude           43000.0                      36.856697   
Longitude          43000.0                     -95.939067   
Income             43170.0                   93227.389622   
Kidhome            43170.0                       1.004749   
Teenhome           43170.0                       0.998633   
Dt_Customer          43170  2013-05-11 22:09:47.407922176   
NumWebVisitsMonth  43170.0                       16.59129   

                                   min                  25%  \
Age                               17.0                 32.0   
Duration                           0.0                102.0   
Campaign                           1.0                  1.0   
Previous                           0.0                  0.0   
Emp.var.rate                      -3.4                 -1.8   
Cons.price.idx                  92.201               93.075   
Cons.conf.idx                    -50.8                -42.7   
Euribor3m                        0.634                1.344   
Nr.employed                     4964.0               5099.0   
Date               2015-01-01 00:00:00  2016-04-04 00:00:00   
Latitude                        24.396             30.61475   
Longitude                     -124.997           -110.49425   
Income                          5841.0              49608.0   
Kidhome                            0.0                  0.0   
Teenhome                           0.0                  0.0   
Dt_Customer        2012-01-01 00:00:00  2012-07-16 00:00:00   
NumWebVisitsMonth                  1.0                  9.0   

                                   50%                  75%  \
Age                               38.0                 47.0   
Duration                         179.0                319.0   
Campaign                           2.0                  3.0   
Previous                           0.0                  0.0   
Emp.var.rate                       1.1                  1.4   
Cons.price.idx                  93.749               93.994   
Cons.conf.idx                    -41.8                -36.4   
Euribor3m                        4.857                4.961   
Nr.employed                     5191.0               5228.0   
Date               2017-07-04 00:00:00  2018-09-30 00:00:00   
Latitude                        36.761             43.11325   
Longitude                     -95.8995            -81.42775   
Income                         93009.5             136740.5   
Kidhome                            1.0                  2.0   
Teenhome                           1.0                  2.0   
Dt_Customer        2013-03-01 00:00:00  2014-03-28 00:00:00   
NumWebVisitsMonth                 17.0                 25.0   

                                   max           std  
Age                               98.0     10.437957  
Duration                        4918.0    258.666033  
Campaign                          56.0      2.772294  
Previous                           7.0      0.497366  
Emp.var.rate                       1.4      1.573898  
Cons.price.idx                  94.767      0.579548  
Cons.conf.idx                    -26.9      4.637186  
Euribor3m                        5.045      1.737117  
Nr.employed                     5228.0     72.337206  
Date               2019-12-31 00:00:00           NaN  
Latitude                        49.384      7.2

In [94]:
df_bank_merge.to_csv('../data/eda_limpieza_files/bank_merge.csv')
